Civi/Core/CiviEventDispatcher.php

```diff
@@ -126,15 +144,53 @@ public function dispatch($eventName, Event $event = NULL) {
       }
     }
     $this->bindPatterns($eventName);
-    return parent::dispatch($eventName, $event);
+    if (version_compare(Kernel::VERSION, '4.3', '>=')) {
+      return $this->dispatcher->dispatch($event, $eventName);
```

Comment:
@demeritcowboy pointed to a backtrace for an error here on D9 (presumably Symfony 4.3+). It was when running `civiExit()`, ie

```php
Civi::dispatcher()->dispatch('civi.core.exit');
```

And the main msg was:

```
TypeError: Argument 1 passed to "Symfony\Component\EventDispatcher\EventDispatcherInterface::dispatch()" must be an object, "NULL" given. in Symfony\Component\EventDispatcher\EventDispatcher->dispatch() (line 63 of ...\vendor\symfony\event-dispatcher\EventDispatcher.php).
```

I guess the issue is that they're pushing for some kind of object. Could use `stdClass` as a placeholder?

```suggestion
        return $this->dispatcher->dispatch($event ?: new stdClass(), $eventName);
```

Reply:
or even new Event() and just use the Event Class

Reply:
Couldn't we just have a conditional to ensure `$event` is defined - if it's null there's no point dispatching anything anyway?

Reply:
> (@MikeyMJCO) if it's null there's no point dispatching anything anyway

Depends if you're talking Symfony 2/3-style or PSR-14-style. (Symfony 4.3/5 is a hybrid that can do both styles.)

In PSR-14-style (`distpach($object)`), I agree there's no point in `$object===NULL` - you *must* have an object to signal the type of the event.

In Symfony 2/3-style, `dispatch('civi.core.exit')` or `dispatch('civi.core.exit', NULL)` are similar to `module_invoke_all('civicrm_exit')`. They tell you that something happened - but no extra data is available/sensible. (*Maybe if there was an exit-code or exit-reason, then you might signal that. 🤷 *)

> (@seamuslee001) or even new Event() and just use the Event Class

FWIW, `Event` is deprecated in v4.3. However, they do add a replacement `GenericEvent` - which is maybe closer to Civi's `GenericHookEvent` (except that `GenericHookEvent` explicitly supports field-orderings and pass-by-reference; not sure if `GenericEvent` does that).

(Note: I had thought about `stdClass` because it's available very early - even before the classloader is setup. But we're not gonna be dispatching events at such an early stage. So it's OK to use a richer class.)

But yes - for a data-less event, you could use `GenericEvent` or `GenericHookEvent` instead of `stdClass`.

Reply:
So I think https://github.com/symfony/contracts/blob/main/EventDispatcher/Event.php is basically what the Event Class is in Symfony 2/3 and I would note that GenericEvent extends that so I would probably lean towards Generic Hook Event just to keep it all CiviCRM and then all we need to do is update GenericEvent

Reply:
@seamuslee001 Yeah, I agree `GenericHookEvent` seems the better bet -- keep things more consistent.

Reply:
Agreed :-)